File: src/video_client.rs
use crate::eventbroker::{EventBroker, timestamp_now};
use crate::error::{Result};
use crate::protobuf:: {
    PurpleDropEvent,
    purple_drop_event::Msg,
    Image,
    ImageTransform,
};

use log::*;

use serde::Deserialize;

use std::net::SocketAddr;
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;

struct State {
    connected: bool,
    last_frame: i32,
}

/// Creates thread to monitor video server (pdcam) for new frames and transforms,
/// and publishes data to event broker. 
///
/// If it fails to connect, it will simply wait for a timeout period and try again. 
#[derive(Clone)]
pub struct VideoClient {
    eventbroker: Arc<Mutex<EventBroker>>,
    host: SocketAddr,
    state: Arc<Mutex<State>>,
}

#[derive(Clone, Debug, Deserialize)]
struct TransformResponse {
    transform: Option<Vec<Vec<f32>>>,
    image_width: i32,
    image_height: i32,
}

fn update(host: SocketAddr, state: &mut State, eventbroker: &mut EventBroker, http: &reqwest::blocking::Client) -> Result<()> {
    let resp = http.get(&format!("http://{}/transform", host.to_string())).send()?;
    let transform_resp: TransformResponse = resp.json()?;
    let mut resp = http.get(&format!("http://{}/latest", host.to_string()))
        .header("X-Min-Frame-Number", (state.last_frame + 1).to_string())
        .send()?;
    state.last_frame = resp.headers().get("X-Frame-Number").unwrap().to_str()?.parse()?;
    debug!("Received frame {}", state.last_frame);
    let mut jpeg: Vec<u8> = vec![];
    resp.copy_to(&mut jpeg)?;

    // Build and send ImageTransform event
    let transform = match transform_resp.transform {
        Some(t) => t.concat(),
        None => vec![]
    };
    // TODO: Get QR codes from server and return
    let msg = ImageTransform{
        timestamp: Some(timestamp_now()),
        transform: transform,
        image_width: transform_resp.image_width,
        image_height: transform_resp.image_height,
        qr_codes: vec![],
    };
    let event = PurpleDropEvent{msg: Some(Msg::ImageTransform(msg))};
    eventbroker.send(event);

    // Build and send image event
    let msg = Image {
        timestamp: Some(timestamp_now()),
        image_data: jpeg,
    };
    let event = PurpleDropEvent{msg: Some(Msg::Image(msg))};
    eventbroker.send(event);
    Ok(())
}

fn client_thread(host: SocketAddr, state: Arc<Mutex<State>>, eventbroker: Arc<Mutex<EventBroker>>) {
    let http = reqwest::blocking::Client::new();
    loop {
        let result;
        {
            let mut state = state.lock().unwrap();
            let mut eventbroker = eventbroker.lock().unwrap();
            result = update(host, &mut *state, &mut *eventbroker, &http);
        }
        
        // on success, request again immediately
        // on error, hold off for a period of time
        match result {
            Ok(_) => (),
            Err(e) => {
                warn!("Failed to retrieve video info: {:?}", e);
                // If we fail to read, reset the frame counter to 0. 
                // If the server restarted, the frame counter will have too,
                // and if it didn't, we just might get a duplicate frame
                let mut state = state.lock().unwrap();
                state.last_frame = 0;
                thread::sleep(Duration::from_millis(5000));
            },
        }
    }
}

impl VideoClient {
    pub fn new(eventbroker: EventBroker, host: SocketAddr) -> VideoClient {
        VideoClient{
            eventbroker: Arc::new(Mutex::new(eventbroker.clone())),
            host,
            state: Arc::new(Mutex::new(State{connected: false, last_frame: 0}))}
    }

    pub fn start(&self) {
        let cloned_state = self.state.clone();
        let cloned_host = self.host.clone();
        let cloned_broker = self.eventbroker.clone();
        //let mut cloned_client = Box::new(self.clone());
        thread::spawn(move || {
            client_thread(cloned_host, cloned_state, cloned_broker);
        });
    }
}

File: purpledrop/video_client.py

import logging
import requests
import threading
import time
from typing import AnyStr

import purpledrop.protobuf.messages_pb2 as messages_pb2

class VideoClient(object):
    def __init__(self, host: str, callback=None):
        self.host = host
        self.callback = callback
        self.last_frame = 0
        self.connected = False

        self.thread = threading.Thread(name="VideoClient", daemon=True, target=self.run)
        self.thread.start()

    def register_callback(self, callback):
        self.callback = callback

    def on_callback(self, *args, **kwargs):
        if self.callback is not None:
            self.callback(*args, **kwargs)

    def run(self):
        while True:
            self.last_frame = 0
            s = requests.Session()

            try:
                resp = requests.get(
                    f"http://{self.host}/latest",
                    headers={'X-Min-Frame-Number': str(self.last_frame + 1)},
                    stream=True,
                )
                resp.raise_for_status()
                self.last_frame = int(resp.headers.get('X-Frame-Number'))
                jpeg_bytes = resp.content

                resp = requests.get(f"http://{self.host}/transform")
                resp.raise_for_status()
                transform = resp.json()

                self.on_callback(jpeg_bytes, transform)
                if not self.connected:
                    self.connected = True
                    logging.info("Connected to video host")

            except requests.exceptions.RequestException as ex:
                logging.debug(f"Failed to capture frame from video host: {ex}")
                if(self.connected):
                    self.connected = False
                    logging.info("Lost connection to video host")
                time.sleep(3.0)


class VideoClientProtobuf(object):
    def __init__(self, host, callback=None):
        self.client = VideoClient(host, self.handler)
        self.callback = callback

    def register_callback(self, callback):
        self.callback = callback

    def on_callback(self, *args, **kwargs):
        if self.callback is not None:
            self.callback(*args, **kwargs)

    def handler(self, jpeg, transform):
        image_event = messages_pb2.PurpleDropEvent()
        transform_event = messages_pb2.PurpleDropEvent()

        timestamp = time.time()
        image_event.image.timestamp.seconds = int(timestamp)
        image_event.image.timestamp.nanos = int((timestamp % 1) * 1e9)
        image_event.image.image_data = jpeg
        transform_event.image_transform.timestamp.seconds = int(timestamp)
        transform_event.image_transform.timestamp.nanos = int((timestamp % 1) * 1e9)
        if transform['transform'] is not None:
            transform_event.image_transform.transform[:] = sum(transform['transform'], [])
        else:
            transform_event.image_transform.transform[:] = []
        transform_event.image_transform.image_width = transform['image_width']
        transform_event.image_transform.image_height = transform['image_height']
        self.on_callback(image_event, transform_event)

